File: material/quad/quad4.c
#include <stdio.h>
#include <math.h>

int quad(float a, float b, float c, float *r1, float *r2)
{
    float det = b * b - 4.0f * a * c;
    if(det < 0.0f)
        return 0;
    *r1 = (-b + sqrt(det)) / (2.0f * a);
    if(det == 0.0f)
        return 1;
    *r2 = (-b - sqrt(det)) / (2.0f * a);
    return 2;
}

int main(void)
{
    FILE *fp = fopen("test.txt", "r");
    if(!fp) {
        fprintf(stderr, "File not found\n");
        return 1;
    }
    float largest_root = -1.0f;
    float largest_a = -1.0f;
    float largest_b = -1.0f;
    float largest_c = -1.0f;
    for(int i = 0; i < 10000; i++) {
        float a;
        float b;
        float c;
        int num_read = fscanf(fp, "%f %f %f", &a, &b, &c);
        if(num_read != 3) {
            fprintf(stderr, "Unable to read values: could only read %d\n", num_read);
            fclose(fp);
            return 1;
        }
        float r1, r2;
        int num_roots = quad(a, b, c, &r1, &r2);
        if(num_roots == 2) {
            if(r2 > largest_root) {
                largest_root = r2;
                largest_a = a;
                largest_b = b;
                largest_c = c;
            }
        }
        if(num_roots != 0) {
            if(r1 > largest_root) {
                largest_root = r1;
                largest_a = a;
                largest_b = b;
                largest_c = c;
            }
        }
    }
    fclose(fp);
    printf("Largest root is %.2f. The function is (%.3f, %.3f, %.3f).\n",
            largest_root, largest_a, largest_b, largest_c);
}

File: material/quad/quad4.py
import math
import sys

def quad(a, b, c):
    det = b * b - 4.0 * a * c
    if det < 0:
        return []
    r1 = (-b + math.sqrt(det)) / (2.0 * a)
    if det == 0:
        return [r1]
    else:
        r2 = (-b - math.sqrt(det)) / (2.0 * a)
        return [r1, r2]

max_root = -1
max_a = -1
max_b = -1
max_c = -1
with open('test.txt', 'r') as f:
    for line in f:
        words = line.split()
        numbers = list()
        for word in words:
            numbers.append(float(word))
        if len(numbers) != 3:
            print >> sys.stderr, 'Could not parse'
        else:
            roots = quad(numbers[0], numbers[1], numbers[2])
            for root in roots:
                if root > max_root:
                    max_root = root
                    max_a = numbers[0]
                    max_b = numbers[1]
                    max_c = numbers[2]

    print 'Maximum root is %.2f for equation (%.3f, %.3f, %.3f)' % (max_root, max_a, max_b, max_c)

